parsers/yaml/parse.cc

```diff
@@ -0,0 +1,851 @@
+          [](auto& m) {
+            m.push(TagDirective);
+            m.add(TagPrefix, 1);
+            m.add(TagHandle, 2);
+            m.term();
+            m.pop(TagDirective);
+          },
+
+        R"(%[[:alpha:]]+[^#\r\n]*)" >>
+          [](auto& m) {
+            std::cerr << "Unknown directive: " << m.match(1).view()
+                      << std::endl;
+            m.add(UnknownDirective, 1);
+          },
+
+        R"([ \t]*\.\.\.(?:\r?\n| )+)" >> [](auto&) { return; },
+
+        R"([ \t]*(---)([ \t]+))" >>
+          [](auto& m) {
+            m.push(Document);
+            m.add(DocumentStart, 1);
+            m.add(Whitespace, 2);
+            m.mode("document");
+          },
+
+        R"([ \t]*(---)(\r?\n))" >>
+          [](auto& m) {
+            m.push(Document);
+            m.add(DocumentStart, 1);
+            m.add(NewLine, 2);
+            m.mode("document");
+          },
+
+        // If we reach this point, then there was no preamble and what follows
+        // is the document itself
+        "^" >>
+          [](auto& m) {
+            m.push(Document);
+            m.mode("document");
+          },
+      });
+
+    // Every stream is made up of zero or more documents.  Document
+    // mode is the mode most would think of as "YAML", but can contain
+    // JSON-like flow documents, which are handled in a separate mode.
+    p("document",
+      {
+        R"(([ \t]*)(#[^\r\n]*))" >>
+          [](auto& m) {
+            if (m.match(1).len > 0)
+            {
+              m.add(Whitespace, 1);
+            }
+            m.add(Comment, 2);
+          },
+
+        R"([ \t]+)" >> [](auto& m) { m.add(Whitespace); },
+
+        R"((\r?\n)(#[^\r\n]*))" >>
+          [](auto& m) {
+            m.add(NewLine, 1);
+            m.add(Comment, 2);
+          },
+
+        R"(\r?\n)" >> [](auto& m) { m.add(NewLine); },
+
+        // text that looks like a directive in a document
+        // %[[:alpha:]]+ : A percent sign followed by one or more alphabetic
+        // characters
+        // (?:[ \t]+[^\s]+) : One or more spaces or tabs followed by one or more
+        // characters which are NOT whitespace
+        // ([ \t]+#[^\r\n]*)? : Zero or one spaces or tabs followed by a comment
+        // (optional)
+        R"((%[[:alpha:]]+(?:[ \t]+[^\s]+))([ \t]+#[^\r\n]*)?)" >>
+          [](auto& m) {
+            m.add(MaybeDirective, 1);
+            if (m.match(2).len > 0)
+            {
+              m.add(Comment, 2);
+            }
+          },
+
+        R"((---)(\r?\n))" >>
+          [](auto& m) {
+            m.term();
+            m.pop(Document);
+            m.push(Document);
+            m.add(DocumentStart, 1);
+            m.add(NewLine, 2);
+          },
+
+        R"((---)([ \t]+))" >>
+          [](auto& m) {
+            m.term();
+            m.pop(Document);
+            m.push(Document);
+            m.add(DocumentStart, 1);
+            m.add(Whitespace, 2);
+          },
```

Comment:
I wonder if this would be clearer as 
```
        R"((---)([ \t]*))(\r?\n))" >>
          [](auto& m) {
            m.term();
            m.pop(Document);
            m.push(Document);
            m.add(DocumentStart, 1);
            if (m.match(2).len > 0)
              m.add(Whitespace, 2);
            m.add(NewLine, 3);
          },
```
Can you have comments on the start a document line?